text/0000-alloca.md

```diff
@@ -0,0 +1,147 @@
+we invite the kernel to swap in a different process instead, which is very probably the major cause of the slowdown
+here).
+
+Still, even with the flaws in this benchmark,
+[The Myth of RAM](http://www.ilikebigbits.com/blog/2014/4/21/the-myth-of-ram-part-i) argues quite convincingly for the
+benefits of memory frugality.
+
+
+# Detailed design
+[design]: #detailed-design
+
+The standard library function can simply `panic!(..)` within the `reserve(_)` method, as it will be replaced when
+translating to MIR. The `StackSlice` type can be implemented as follows:
+
+```Rust
+/// A slice of data on the stack
+pub struct StackSlice<'a, T: 'a> {
+    slice: &'a [T],
+}
+
+impl<'a, T: 'a> Deref for StackSlice<'a, T> {
+    type Target = [T];
+
+    fn deref(&self) -> &[T] {
+        return self.slice;
+    }
+}
+```
+
+`StackSlice`'s embedded lifetime ensures that the stack allocation may never leave its scope. Thus the borrow checker
+can uphold the contract that LLVM's `alloca` requires.
+
+MIR Level: We need a way to represent the dynamic stack `alloca`tion with both the number of elements and the concrete
+type of elements. Then while building the MIR, we need to replace the `Calls` from HIR with it.
+
+Low-level: LLVM has the `alloca` instruction to allocate memory on the stack. We simply need to extend trans to emit it
+with a dynamic `<NumElements>` argument when encountering the aforementioned MIR.
+
+With a LLVM extension to un-allocate the stack slice we could even restrict the stack space reservation to the lifetime
+of the allocated value, thus increasing locality over C code that uses alloca (which so far is suboptimally implemented
+by some compilers, especially with regards to inlining).
+
+# How to teach this
+
+Add the following documentation to libcore:
+
+```
+*** WARNING *** stay away from this feature unless you absolutely need it.
+Using it will destroy your ability to statically reason about stack size.
+
+Apart from that, this works much like an unboxed array, except the size is
+determined at runtime. Since the memory resides on the stack, be careful
+not to exceed the stack limit (which depends on your operating system),
+otherwise the resulting stack overflow will at best kill your program. You
+have been warned.
+
+Valid uses for this is mostly within embedded system without heap allocation.
+```
+
+Also add an example (perhaps a sort algorithm that uses some scratch space that will be heap-allocacted with `std` and
+stack-allocated with `#[no_std]` (noting that the function would not be available on no-std systems at all were it not
+for this feature).
+
+Do not `pub use` it from `std::mem` to drive the point home.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Even more stack usage means the dreaded stack limit will probably be reached even sooner. Overflowing the stack space
+leads to segfaults at best and undefined behavior at worst. On unices, the stack can usually be extended at runtime,
```

Comment:
I am working on integrating stack probes, which are required for robustness on our embedded system ([ARTIQ](m-labs.hk/artiq/)). I expect to implement them in a way generic enough to be used on all bare-metal platforms, including no-MMU, as well as allow for easy shimming on currently unsupported OSes.